### DISCUSSION TITLE
Adding back and forward button functionality

### DISCUSSION THREAD

**Discussion by Abhay0294** [eclipse-theia/theia#11278] [Q&A] [answered]
How can I add the back and forward button functionality at perspective level.

**Requirement:**
1. I'll be having 2 buttons on the top panel one for the back and the other for the forward functionality.
2. The back and forward functionality should work  at the perspective(group of widgets visible on a screen at a time) level.
3. It means the functionality should be able to capture/store the previously as well as the currently opened perspective, so that on the click of back or forward button we can toggle between the previously and currently opened perspective.

Requesting all genius community members to provide the relevant suggestions for the implementation.

**Comment by vince-fugnitto** [comment, marked as answer]
Thank you for the discussion!

The `@theia/toolbar` extension contributes a global toolbar with the `back` and `forward` functionality you are looking for. You could use the functionality already, enhance it, or use it as an example to contribute your own.

![image](https://user-images.githubusercontent.com/40359487/172839038-3fd1d30a-77e9-439b-aa70-a7b5d2c733a1.png)

It currently uses the editor navigation history. 

https://github.com/eclipse-theia/theia/blob/b2e44a9367b1d9c9b1a0977edf57deab3cc50759/packages/editor/src/browser/editor-command.ts#L89-L104

https://github.com/eclipse-theia/theia/blob/b2e44a9367b1d9c9b1a0977edf57deab3cc50759/packages/editor/src/browser/navigation/navigation-location-service.ts#L26-L31

I'm not exactly sure what you mean by perspective (ex: if you want to restore views and widgets), but that will likely require a more significant change where you would need to keep track of everything in your custom history.
